sdk/python/lib/pulumi/resource.py

```diff
@@ -517,18 +521,22 @@ def _expand_providers(options: 'ResourceOptions'):
 
 def _collapse_providers(opts: 'ResourceOptions'):
     # If we have only 0-1 providers, then merge that back down to the .provider field.
-    providers: List['ProviderResource'] = opts.providers
+    providers: Union[Mapping[str, ProviderResource], List[ProviderResource], None] = opts.providers
```

Comment:
is there a difference between `Optional[Union[A, B]]` and `Union[A, B, None]`?

Reply:
No difference. The former is the type hint that mypy spits out. Updated to the more readable version.